src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -106,4 +121,287 @@ object RuleUtils {
+        val curFileSet = location.allFiles
+          .map(f => FileInfo(f.getPath.toString, f.getLen, f.getModificationTime))
+        filesAppended =
+          curFileSet.filterNot(index.allSourceFileInfos.contains).map(f => new Path(f.name))
+        // TODO: Hybrid Scan delete support.
+
+        // If BucketSpec of index data isn't used (e.g., in the case of FilterIndex currently),
+        // or the source format is not parquet. We could read appended data from source files
+        // along with the index data.
+        val readPaths = {
+          if (useBucketSpec || !isParquetSourceFormat) {
+            // FilterIndexRule: Since the index data is in "parquet" format, we cannot read
+            // source files in formats other than "parquet" using 1 FileScan node as the
+            // operator requires files in one homogenous format. To address this, we need
+            // to read the appended source files using another FileScan node injected into
+            // the plan and subsequently merge the data into the index data. Please refer below
+            // [[Union]] operation.
+            index.content.files
+          } else {
+            val files = index.content.files ++ filesAppended
+            // Reset filesAppended as it's handled with index data.
+            filesAppended = Nil
+            files
+          }
+        }
+
+        val newLocation = new InMemoryFileIndex(spark, readPaths, Map(), None)
+        val relation = HadoopFsRelation(
+          newLocation,
+          new StructType(),
+          StructType(index.schema.filter(baseRelation.schema.contains(_))),
+          if (useBucketSpec) Some(index.bucketSpec) else None,
+          new ParquetFileFormat,
+          Map())(spark)
+
+        val updatedOutput =
+          baseOutput.filter(attr => relation.schema.fieldNames.contains(attr.name))
+        baseRelation.copy(relation = relation, output = updatedOutput)
+    }
+
+    if (filesAppended.nonEmpty) {
+      // If there are unhandled appended files, we need to create additional plans
+      // by the following steps:
+      // Step 1) Generate a plan(planForAppended) from the current plan to read
+      //   the appended files similar to indexPlan.
+      // Step 2) If Shuffle is required, perform shuffle for the plan.
+      // Step 3) Merge both indexPlan and planForAppended by using [[BucketUnion]] or [[Union]].
+      // For more details, see https://github.com/microsoft/hyperspace/issues/150.
+
+      val planForAppended =
+        transformPlanToReadFromAppendedFiles(spark, index.schema, plan, filesAppended)
+      if (useBucketSpec) {
+        // If Bucketing information of the index is used to read the index data, we need to
+        // shuffle the appended data in the same way to correctly merge with bucketed index data.
+
+        // Clear sortColumnNames as BucketUnion does not keep the sort order within a bucket.
+        val bucketSpec = index.bucketSpec.copy(sortColumnNames = Seq())
```

Comment:
nit: we can use `Nil` here. (more consistent in this file)

Reply:
Maybe we should just use `numBuckets` in `BucketUnion*`? Are we using any other fields from `BucketSpec`?

Also, can you add an assert in `BucketUnionExec.outputPartitioning` that the number of partitions is same as the number of buckets? I think that's our assumption right?
```
assert(children.head.outputPartitioning.asInstanceOf[HashPartitioning].numPartitions == bucketSpec.numBuckets)
```

Reply:
Btw, do we need to check if the bucketed columns are the partitioning expressions?

Reply:
Also, as far as I know, bucket spec is meant to be used for the scan node. So it may be more confusing if we print out it in the `BucketUnion`. We can take this up as a separate PR, but I think we need to address this.

Reply:
Ok I revised the comment a bit. But I think the information might help to understand the behavior? Does this look confusing?
```
 +- *(5) Project [o_orderkey#61L, o_orderdate#65, o_shippriority#68]
:        +- *(5) SortMergeJoin [c_custkey#5L], [o_custkey#62L], Inner
:           :- *(1) Project [c_custkey#5L]
:           :  +- *(1) Filter ((isnotnull(c_mktsegment#11) && (c_mktsegment#11 = BUILDING)) && isnotnull(c_custkey#5L))
:           :     +- *(1) FileScan parquet [c_custkey#5L,c_mktsegment#11] Batched: true, Format: Parquet, Location: InMemoryFileIndex[wasb://eunjin-hyperspace-test@taruntestdiag.blob.core.windows.net/indexes/index..., PartitionFilters: [], PushedFilters: [IsNotNull(c_mktsegment), EqualTo(c_mktsegment,BUILDING), IsNotNull(c_custkey)], ReadSchema: struct<c_custkey:bigint,c_mktsegment:string>, SelectedBucketsCount: 200 out of 200
:           +- *(4) Sort [o_custkey#62L ASC NULLS FIRST], false, 0
:              +- BucketUnion 200 buckets, bucket columns: [o_custkey]             <===============
:                 :- *(2) Project [o_orderkey#61L, o_custkey#62L, o_orderdate#65, o_shippriority#68]
:                 :  +- *(2) Filter (((isnotnull(o_orderdate#65) && (o_orderdate#65 < 1995-03-15)) && isnotnull(o_custkey#62L)) && isnotnull(o_orderkey#61L))
:                 :     +- *(2) FileScan parquet [o_custkey#62L,o_orderkey#61L,o_orderdate#65,o_shippriority#68] Batched: true, Format: Parquet, Location: InMemoryFileIndex[wasb://eunjin-hyperspace-test@taruntestdiag.blob.core.windows.net/indexes/index..., PartitionFilters: [], PushedFilters: [IsNotNull(o_orderdate), LessThan(o_orderdate,1995-03-15), IsNotNull(o_custkey), IsNotNull(o_orde..., ReadSchema: struct<o_custkey:bigint,o_orderkey:bigint,o_orderdate:string,o_shippriority:int>, SelectedBucketsCount: 200 out of 200
:                 +- Exchange hashpartitioning(o_custkey#62L, 200)
:                    +- *(3) Project [o_orderkey#61L, o_custkey#62L, o_orderdate#65, o_shippriority#68]
:                       +- *(3) Filter (((isnotnull(o_orderdate#65) && (o_orderdate#65 < 1995-03-15)) && isnotnull(o_custkey#62L)) && isnotnull(o_orderkey#61L))
:                          +- *(3) FileScan parquet [o_custkey#62L,o_orderkey#61L,o_orderdate#65,o_shippriority#68] Batched: true, Format: Parquet, Location: InMemoryFileIndex[wasb://eunjin-hyperspace-test@taruntestdiag.blob.core.windows.net/data/tpch-par..., PartitionFilters: [], PushedFilters: [IsNotNull(o_orderdate), LessThan(o_orderdate,1995-03-15), IsNotNull(o_custkey), IsNotNull(o_orde..., ReadSchema: struct<o_custkey:bigint,o_orderkey:bigint,o_orderdate:string,o_shippriority:int>
```

Reply:
I think it may be better if we print out the output partitioning, similar to `Exchange` since we are unioning plans that have the same hash partitioning.

Reply:
But new comment sounds better and let's take this up separately.